### DISCUSSION TITLE
pytest 'hangs' upon completing test run

### DISCUSSION THREAD

**Discussion by JohnStokes228** [dagster-io/dagster#18495] [Q&A]
hi all,

i've hit an issue attempting to run pytest for dagster assets. it appears that after running a successful test, pytest simply hangs and never returns control to the console back to me.

a minimal example is:

```
@asset(name='hello_world')
def arbitrary_asset():
    Output(value=12)


def test_arbitrary_asset():
    assert arbitrary_asset() == 12
```

my environment looks like:

dependencies:
  - python=3.9.17
  - pip=23.1.2
  - pip:
      - pins==0.8.3
      - sqlfluff==2.3.1
      - black==22.3.0
      - pluggy==1.3.0
      - xlrd==1.2.0
      - openpyxl==3.1.2
      - jupyter
      - pre-commit==3.4.0
      - pyodbc==4.0.39
      - scikit-learn==1.3.0
      - pyprojroot==0.3.0
      - pandas==2.1.0
      - pyarrow==13.0.0
      - numpy==1.24.4
      - matplotlib==3.7.2
      - seaborn==0.12.2
      - lightgbm==4.1.0
      - shap==0.42.1
      - optuna==3.3.0
      - imbalanced-learn==0.11.0
      - jupyter-contrib-nbextensions
      - tubular==1.0.0
      - jupytext==1.15.2
      - gitpython==3.1.40
      - dagster==1.5.3
      - dagster-webserver==1.5.3
      - dagster-pandas==0.21.3
      - pytest==7.4.3

run on a Windows Server 2019 machine

potentially relevant note, the problem isnt seen if the tests fail. Any guidance on what I might be setting up wrong would be welcome as there isnt any trace to debug from.

cheers!

**Comment by JohnStokes228** [comment]
as an update to this:
- the issue isnt seen in the identical environment when run on linux machines
- the issue isnt seen on windows machines when running pytest with the xdist plugin (? why is this the case?)
looks to be some issue in the way its dealing with the async - potentially GIL related since it works on linux?

**Reply by Lord-Kamina** [reply]
Was this ever figured out? I'm seeing a similar issue and wonder whether they're related

**Reply by BQTHake003** [reply]
I have the same issue. Whenever pytest finishes, it hangs indefinitely and the control never returns to the console. 

And I am using a fixture and tried to play with the scopes but with no difference. I am also using pytest-bdd plugin

I am using Windows 10
python 3.11
pytest 7.4.4

**Reply by JohnStokes228** [reply]
unfortunately not no, for our use case we opted to simply enable xdist to get past the issues we were having as we had no leads on what to fix. be happy to help with any testing etc... if anyone had any ideas though

**Reply by BQTHake003** [reply]
@JohnStokes228 thanks for your quick reply.

It turned out that I wasn't clearning out resources correctly on one of my objects that was running. This object is threaded object and I was counting on __del__ to stop the thread and destroy the object, but this wasn't the case.

So I added a function within the object to properly stop the thread when needed and clearing the object resources before deleting it and the issue was fixed.

It wasn't pytest. I haven't installed xdist though.

**Reply by streamnsight** [reply]
in my experience, when I had the same problem, the reason was the order of my patches:

let's say a module does a long init (like a DB connection check that may fail in testing because offline)
if the module is imported in another module, but is not used in the test. The module is then patched, the test run won't show failure of the module init.

In my case I had patched a module, but it was already being imported by another module before the patch. It wasn't being used in that particular module, so failure went silent, but made the pytest hang for some time at the end of the run.

By checking where modules were imported and where they are patched, I re-ordered my patches and eliminated this issue. 

Remember that patches are applied from the function name and up in reverse order since they are wrappers around the previous patch and finally around the function...

I hope this helps others with the same issue.
